swaywm/wlroots
# Output damage tracking

Roadmap:

- [x] Do not actually swap buffers when nothing changed - sleep and retry later
- [x] Use surface damage to populate output damage
- [x] Only draw affected regions
- [x] Don't render the whole output when software cursor moves (#572)
- [x] Double buffering artifacts
- [x] Software cursor artifacts when changing cursor image (try chromium, evince on borders or over text)
- [x] Shadow artifacts when resizing a window (try weston-terminal, also happens when un-maximizing a view)
- [x] Listen to subsurface commit event
- [x] Listen to xdg popup commit event
- [x] Listen to wl popup commit event
- [x] subsurface artifacts when animated (try gnome-calculator)
- [x] Fullscreen views
- [x] Xwayland popups not visible in fullscreen
- [x] Damage whole output on mode change (try resizing wayland backend window)
- [x] Fix software cursors being drawn multiple times outside the damage zone
- [x] Fix hardware cursor not moving when not swapping buffers in DRM backend (with atomic interface)
- [x] Fix buffers too old in X11 backend
- [x] Handle drag icons
- [x] Fix decorations when resizing views
- [x] Fix QT decorations (try qtcreator)
- [x] Output scale
- [x] Output transforms
- [x] View rotation
- [x] Wait for damage before trying to render (with rate limiting)
- [x] Use `EGL_EXT_buffer_age` to make sure we can re-use old buffers (if not, damage the whole output) and to support triple buffering
- [x] Wrap GL calls in the renderer
- [x] Update examples
- [x] Fix the screenshooter blocking when there's no damage
- [x] When switching to another tty and back, one of the buffers gets cleared to black
- [x] Fix DRM with multiple outputs
- [x] Fix mpv not refreshing video when moving software cursor on Wayland backend
- [x] Fix DRM modesetting glitches (maybe remove `retry_pageflip`?)
- [x] Changing views z-index doesn't damage them
- [x] Fix urxvt carret damage on output with scale=2
- [x] Fix artifacts when unmaximizing/leaving fullscreen for rotated views
- [x] Fix artifacts when entering fullscreen for mpv between two outputs
- [ ] Consider integrating some rootston stuff into wlroots

If you want to show which parts of the screen are repainted by damage tracking, add this in `rootston/output.c` on line 421:

```c
wlr_renderer_clear(output->desktop->server->renderer, 1, 1, 1, 1);
```

Test plan: this is a test matrix
* Try all backends
* Try all output transforms and change output scale (integral, fractional)
* Try all shells
* Try subsurfaces, popups, drag icons
* Try cursors
* Try fullscreen surfaces
* Try view rotation

Future work:
- When damaging a whole view child, do not damage the whole parent
- Use opaque regions to do not always draw the full stack of views (#277)
- Provide damage to backend when swapping buffers, and use `EGL_EXT_swap_buffers_with_damage`/`EGL_KHR_swap_buffers_with_damage` instead of `eglSwapBuffers` if available
- When resuming rendering, figure out where we are in the vblank cycle (see Weston's `start_repaint_loop`). To do so, request a frame event from the backend/output.

Fixes #192
Fixes #572
